cpp/src/arrow/compute/kernels/hash.cc

```diff
@@ -82,21 +82,77 @@ class UniqueAction : public ActionBase {
 
   Status Reserve(const int64_t length) { return Status::OK(); }
 
-  void ObserveNull() {}
+  Status ObserveNull() { return Status::OK(); }
 
   template <class Index>
-  void ObserveFound(Index index) {}
+  Status ObserveFound(Index index) {
+    return Status::OK();
+  }
 
   template <class Index>
-  void ObserveNotFound(Index index) {}
+  Status ObserveNotFound(Index index) {
+    return Status::OK();
+  }
 
   Status Flush(Datum* out) { return Status::OK(); }
 
   std::shared_ptr<DataType> out_type() const { return type_; }
+
+  Status FlushFinal(Datum* out) { return Status::OK(); }
+};
+
+// ----------------------------------------------------------------------
+// Count values implementation (see HashKernel for description of methods)
+
+class ValueCountsAction : ActionBase {
+ public:
+  using ActionBase::ActionBase;
+
+  ValueCountsAction(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : ActionBase(type, pool), count_builder_(pool) {}
+
+  Status Reserve(const int64_t length) {
+    // builder size is independent of input array size.
+    return Status::OK();
+  }
+  Status Reset() {
+    count_builder_.Reset();
+    return Status::OK();
+  }
+
+  // Don't do anything on flush beceause we don't want to finalize the builder
+  // or incur the cost of memory copies.
+  Status Flush(Datum* out) { return Status::OK(); }
+
+  std::shared_ptr<DataType> out_type() const { return type_; }
+
+  // Return the counts corresponding the MemoTable keys.
+  Status FlushFinal(Datum* out) {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(count_builder_.FinishInternal(&result));
+    out->value = std::move(result);
+    return Status::OK();
+  }
+
+  Status ObserveNull() { return Status::OK(); }
```

Comment:
We're going to want to count nulls (perhaps optionally). The null count can be appended at the end

https://issues.apache.org/jira/browse/ARROW-4787